docs/decisions/0021-json-serializable-custom-types.md

```diff
@@ -0,0 +1,163 @@
+
+        if (targetType.GetCustomAttribute<JsonConverterAttribute>() is JsonConverterAttribute tca &&
+            Type.GetType(tca.ConverterType.Name, throwOnError: false) is Type converterType &&
+            Activator.CreateInstance(converterType) is JsonConverter converter)
+        {
+            return converter;
+        }
+
+        return new SimpleJsonConverter();
+    }
+
+    private class JsonConverterWrapper : JsonConverter<object> {
+        private readonly TypeConverter _converter;
+
+        public JsonConverterWrapper(TypeConverter converter)
+        {
+            this._converter = converter;
+        }
+
+        public override object Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            return this._converter.ConvertFromString(null, CultureInfo.InvariantCulture, reader.GetString());
+        }
+
+        public override void Write(Utf8JsonWriter writer, object value, JsonSerializerOptions options)
+        {
+            writer.WriteStringValue(this._converter.ConvertToString(null, CultureInfo.InvariantCulture, value));
+        }
+    }
+
+
+    private class SimpleJsonConverter : JsonConverter<object>
+    {
+        public override object? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            return JsonSerializer.Deserialize(ref reader, typeToConvert, options);
+        }
+
+        public override void Write(Utf8JsonWriter writer, object value, JsonSerializerOptions options)
+        {
+            JsonSerializer.Serialize(writer, value, value.GetType(), options);
+        }
+    }
+
+```
+
+If a developer wants to customize the serialization and/or deserialization of a custom type, they can implement their own `JsonConverter` and apply the `JsonConverterAttribute` their custom type as seen below.
+
+**Example usage of the `JsonConverter`** class and attribute:
+
+```csharp
+    [JsonConverter(typeof(MyCustomTypeConverter))]
+    private sealed class MyCustomType
+    {
+        public int Number { get; set; }
+
+        public string? Text { get; set; }
+    }
+
+    private sealed class MyCustomTypeConverter : JsonConverter<MyCustomType>
+    {
+        public override MyCustomType Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            return JsonSerializer.Deserialize<MyCustomType>(ref reader, options);
+        }
+
+        public override void Write(Utf8JsonWriter writer, MyCustomType value, JsonSerializerOptions options)
+        {
+            JsonSerializer.Serialize(writer, value, options);
+        }
+    }
+```
+
+_When is serialization/deserialization required?_
+
+Required
+
+- **Native to Semantic:** Passing variables from Native to Semantic **will** require serialization of the output of the Native Function from complex type to string so that it can be passed to the LLM.
```

Comment:
If we have an LLM which accepts something other than strings how will this work?

Reply:
can you give an example of this? the assumption here is that semantic functions will always be working with strings, is that not a safe assumption to make?

Reply:
If we have a native function which loads an image and returns a stream and then we want to pass it to a an LLM that can do image to text how would this work?

Reply:
ah ok yes, it would work in the way you've described. image representation came up in the discussion for this ADR, and we agreed that a base64 string representation of the image included in the prompt would be the best approach for this.